File: python/1493.py
# Link: https://leetcode.com/problems/longest-subarray-of-1s-after-deleting-one-element/description/
# Runtime: 399 ms
# Memory: 19.2 MB

class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        left, zeros, longest_subarray = 0, 0, 0
        for right, value in enumerate(nums):
            if value == 0:
                zeros += 1
            
            while zeros > 1:
                if nums[left] == 0:
                    zeros -= 1
                left += 1
            longest_subarray = max(longest_subarray, right - left)
        
        return min(longest_subarray, len(nums) - 1)

File: c++/1493.cpp
// Link: https://leetcode.com/problems/longest-subarray-of-1s-after-deleting-one-element/description/
// Runtime: 36 ms
// Memory: 36.5 MB

class Solution {
public:
  int longestSubarray(const std::vector<int> &nums) {
    int left = 0, zeros = 0, longestSubarray = 0;
    for (int right = 0; right < static_cast<int>(nums.size()); right++) {
      if (nums[right] == 0)
        zeros++;

      while (zeros > 1) {
        if (nums[left] == 0)
          zeros--;
        left++;
      }

      longestSubarray = std::max(longestSubarray, right - left);
    }

    return std::min(longestSubarray, static_cast<int>(nums.size() - 1));
  }
};
